### PR TITLE
Add RFBridge component

### PR DESCRIPTION
**Pull request by jesserockz**
## Description:
Adds code to communicate with the efm8 MCU chip in the Sonoff RFBridge without hardware modifications. 

It is limited to the codes that the RFBridge MCU can decode which mainly included RCSwitch based codes and can learn new protocols with the `learn` function.

This code from [espurna](https://github.com/xoseperez/espurna/blob/dev/code/espurna/rfbridge.ino) was used as a base to get started. 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#433

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
@OttoWinter 
Not sure if this belongs as a component or not, but I created it anyway and I use it. 
Let me know what you think.

File: esphome/components/rf_bridge/__init__.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+RFBridgeData = rf_bridge_ns.struct('RFBridgeData')
+
+CONF_ON_CODE_RECEIVED = 'on_code_received'
+
+CONFIG_SCHEMA = cv.All(cv.Schema({
+    cv.GenerateID(): cv.declare_id(RFBridgeComponent),
+    cv.Optional(CONF_ON_CODE_RECEIVED): automation.validate_automation({
+        cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(RFBridgeReceivedCodeTrigger),
+    }),
+}).extend(uart.UART_DEVICE_SCHEMA))
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/__init__.py]:
```suggestion
}).extend(uart.UART_DEVICE_SCHEMA).extend(COMPONENT_SCHEMA))
```

File: esphome/components/rf_bridge/__init__.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import automation
+from esphome.const import CONF_ID, CONF_TRIGGER_ID, CONF_CODE, CONF_LOW, CONF_SYNC, CONF_HIGH
+from esphome.components import uart
+
+DEPENDENCIES = ['uart']
+
+rf_bridge_ns = cg.esphome_ns.namespace('rf_bridge')
+RFBridgeComponent = rf_bridge_ns.class_('RFBridgeComponent', cg.Component)
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/__init__.py]:
```suggestion
RFBridgeComponent = rf_bridge_ns.class_('RFBridgeComponent', cg.Component, uart.UARTDevice)
```

File: esphome/components/rf_bridge/__init__.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+RFBridgeReceivedCodeTrigger = rf_bridge_ns.class_('RFBridgeReceivedCodeTrigger',
+                                                  automation.Trigger)
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/__init__.py]:
```suggestion
                                                  automation.Trigger.template(RFBridgeData))
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      break;
+    default:
+      ESP_LOGD(TAG, "Unknown action: 0x%02X", action);
+      break;
+  }
+  this->last_ = millis();
+}
+
+void RFBridgeComponent::loop() {
+  static bool RECEIVING = false;
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]: Convert this to a class member variable - I know there's usually just one RFBridge chip hooked up and thus it's ok here. But it's good code style and keeps other integrations where it is an issue from copying this style.

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  if (this->last_ != 0 && millis() - this->last_ > RF_DEBOUNCE) {
+    this->last_ = 0;
+  }
+
+  while (this->available()) {
+    byte c = this->read();
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]:
```suggestion
    uint8_t c = this->read();
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+void RFBridgeComponent::ack_() {
+  ESP_LOGD(TAG, "Sending ACK");
+  this->write(RF_CODE_START);
+  this->write(RF_CODE_ACK);
+  this->write(RF_CODE_STOP);
+  this->flush();
+}
+
+void RFBridgeComponent::decode_() {
+  byte action = uartbuf_[0];
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]:
```suggestion
  uint8_t action = uartbuf_[0];
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+#include "esphome/core/log.h"
+#include <string.h>
+
+namespace esphome {
+namespace rf_bridge {
+
+static const char* TAG = "rf_bridge";
+
+void RFBridgeComponent::ack_() {
+  ESP_LOGD(TAG, "Sending ACK");
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]:
```suggestion
  ESP_LOGV(TAG, "Sending ACK");
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
+#include "rf_bridge.h"
+#include "esphome/core/log.h"
+#include <string.h>
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]:
```suggestion
#include <cstring>
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+void RFBridgeComponent::learn() {
+  ESP_LOGD(TAG, "Learning mode");
+  this->write(RF_CODE_START);
+  this->write(RF_CODE_LEARN);
+  this->write(RF_CODE_STOP);
+}
+
+void RFBridgeComponent::dump_config() { ESP_LOGCONFIG(TAG, "RF_Bridge:"); }
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]: Call `check_uart_settings` in this function. That way you can validate the user has set the right uart baudrate. (see for example tuya/tuya.cpp)

**Review by OttoWinter — COMMENTED** [review]
This is great, thank you! Support for the RFBridge is very appreciated.

The current design is also good.

For my understanding: The workflow for the user is this, right?
 - User enables rf bridge in their config and creates a template switch to turn on learning mode
 - User turns on learning mode and sends the signal they want to learn.
 - Logs print the data, user copies that into their config (for example template switch)

**Comment by jesserockz** [comment]
Thats pretty much spot on. I don't think you can stop learning mode though as it just times out after 30 seconds or when the first code comes in.

I use services as below to switch learn on and to send codes from HA.


```yaml
api:
  services:
    - service: send_rf_code
      variables:
        sync: int
        low: int
        high: int
        code: int
      then:
        - rf_bridge.send_code:
            sync: !lambda 'return sync;'
            low: !lambda 'return low;'
            high: !lambda 'return high;'
            code: !lambda 'return code;'
    - service: learn
      then:
        - rf_bridge.learn
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
+#include "rf_bridge.h"
+#include "esphome/core/log.h"
+#include <cstring>
+
+namespace esphome {
+namespace rf_bridge {
+
+static const char* TAG = "rf_bridge";
```
Comment (OttoWinter) [inline esphome/components/rf_bridge/rf_bridge.cpp]:
```suggestion
static const char *TAG = "rf_bridge";
```

Don't know why travis CI is acting up again, but this is likely required to make format checks pass.

**Review by OttoWinter — COMMENTED** [review]
Code looks good to me, and I'd be happy to merge this when docs are ready :)

**Comment by jesserockz** [comment]
Thanks @OttoWinter.
I have added docs using SIM800L as a template. 
Amazing that I managed to get PR number 433 for them.

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**